### PR TITLE
Add a directory where fluentd configs can be placed for the windows agent to consume.

### PR DESCRIPTION
**Pull request by iantalarico**
This will allow users to expand on the current config

### CODE REVIEW

File: windows-installer/fluent-template.conf [hunk 1]
```diff
@@ -37,8 +37,17 @@
...
+  pos_file WIN_EVT_POS_FILE_PLACE_HOLDER
   # Read every 2 seconds.
   read_interval 2
   tag winevt.raw
 </source>
+
+<source>
+  type tail
+  format none
+  path CUSTOM_LOGS_DIR_PLACE_HOLDER
```
Comment (igorpeshansky) [inline windows-installer/fluent-template.conf]: Hmm, this directive will simply collect logs under the CustomLogs path. We want to give customers a way to augment their logging config instead to add such directives pointing to whatever path they choose.

Let's instead add the following at the top of the config:
```
@include config.d/*.conf
```
That way, users can create files under `$INSTDIR\config.d` with fluentd config directives.

File: windows-installer/setup.nsi [hunk 2]
```diff
@@ -185,13 +189,14 @@ Section "Install"
 
   ; Create a directory to store position files.
   CreateDirectory ${MAIN_INSTDIR}\pos
+  ; Create a directory for custom logs.
+  CreateDirectory ${CUSTOM_LOGS_DIR}
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Let's instead create the `"$INSTDIR\config.d"` directory... You're welcome to parameterize it, but I'm fairly certain it'll never change. Since it's relative to the location of the main config file, maybe parameterize that directory instead, e.g.:
```
; Absolute location of the directory containing the fluentd configuration.
 !define FLUENTD_CONFIG_DIRECTORY "$INSTDIR"

; Absolute location of the fluentd config file.
 !define FLUENTD_CONFIG_LOCATION "$FLUENTD_CONFIG_DIRECTORY\fluent.conf"

; Absolute location of the fluentd config extensions.
 !define FLUENTD_CONFIG_EXTENSIONS "$FLUENTD_CONFIG_DIRECTORY\config.d"
```
...
```
  ; Create a directory for config extensions.
  CreateDirectory ${FLUENTD_CONFIG_EXTENSIONS}
```

File: windows-installer/setup.nsi [hunk 3]
```diff
@@ -79,8 +82,9 @@ RequestExecutionLevel admin
 ; FUNCTIONS
 ;--------------------------------
 
-; Define StrTrimNewLines function to use below.
+; Define functions to use below.
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: This was confusing to me when I first read it. It was clearer before you removed the function name (which implied that the first use of the function with no arguments actually defined it). How about:
```
; Define StrTrimNewLines and StrRep functions to use below.
```
?

File: windows-installer/setup.nsi [hunk 4]
```diff
@@ -212,15 +217,31 @@ Section "Install"
...
       StrCpy $2 "  pos_file '${MAIN_INSTDIR}\pos\winevtlog.pos'$\r$\n"
     ${EndIf}
 
+    ; Look for 'CUSTOM_LOGS_DIR_PLACE_HOLDER', if found replace it with the
+    ; proper path.
+    ${WordFind} "$3" "CUSTOM_LOGS_DIR_PLACE_HOLDER" "#" $4
+    ${If} $4 == "1"
+	  ; Fluentd doesn't properly glob windows paths.  This is a temporary
+	  ; fix until https://github.com/fluent/fluentd/issues/1138 is solved.
+	  ${StrRep} "$5" "${CUSTOM_LOGS_DIR}\**\*" "\" "/"
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Probably irrelevant now...

**Comment by iantalarico** [comment]
Thanks for the feedback, PTAL.

File: windows-installer/setup.nsi [hunk 5]
```diff
@@ -212,15 +217,24 @@ Section "Install"
...
+    ${WordFind} "$3" "WIN_EVT_POS_FILE_PLACE_HOLDER" "#" $4
     ${If} $4 == "1"
       StrCpy $2 "  pos_file '${MAIN_INSTDIR}\pos\winevtlog.pos'$\r$\n"
     ${EndIf}
 
+    ; Look for 'CUSTOM_CONFIG_DIR_PLACE_HOLDER', if found replace it with the
+    ; proper path.
+    ${WordFind} "$3" "CUSTOM_CONFIG_DIR_PLACE_HOLDER" "#" $4
+    ${If} $4 == "1"
+	  ; Fluentd doesn't properly glob windows paths.  This is a temporary
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Most Windows tools I know work just fine with forward slashes instead of backslashes. Would it make sense to just always replace backslashes with forward slashes instead of relying on fluentd to fix this?

File: windows-installer/setup.nsi [hunk 6]
```diff
@@ -212,15 +217,24 @@ Section "Install"
...
-    ; ever be 0 or 1.  We only have it in the template file once.
-    ${WordFind} "$3" "POS_FILE_PLACE_HOLDER" "#" $4
-
-    ; If we hit the place holder line replace it with the proper pos_file.
+	
+    ; Look for 'WIN_EVT_POS_FILE_PLACE_HOLDER', if found replace it with the
+    ; proper pos_file.
+    ${WordFind} "$3" "WIN_EVT_POS_FILE_PLACE_HOLDER" "#" $4
     ${If} $4 == "1"
       StrCpy $2 "  pos_file '${MAIN_INSTDIR}\pos\winevtlog.pos'$\r$\n"
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Now that we have `StrRep`, wouldn't this just become:
```
${StrRep} "$2" "WIN_EVT_POS_FILE_PLACE_HOLDER" "'${MAIN_INSTDIR}\pos\winevtlog.pos'"
```
?

File: windows-installer/setup.nsi [hunk 7]
```diff
@@ -212,15 +217,24 @@ Section "Install"
...
     ${EndIf}
 
+    ; Look for 'CUSTOM_CONFIG_DIR_PLACE_HOLDER', if found replace it with the
+    ; proper path.
+    ${WordFind} "$3" "CUSTOM_CONFIG_DIR_PLACE_HOLDER" "#" $4
+    ${If} $4 == "1"
+	  ; Fluentd doesn't properly glob windows paths.  This is a temporary
+	  ; fix until https://github.com/fluent/fluentd/issues/1138 is solved.
+	  ${StrRep} "$5" "${CUSTOM_CONFIG_DIR}\**\*.conf" "\" "/"
+      StrCpy $2 "@include '$5'$\r$\n"
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Same here: with `StrRep`, isn't this just:
```
${StrRep} "$5" "${CUSTOM_CONFIG_DIR}\**\*.conf" "\" "/"
${StrRep} "$2" "CUSTOM_CONFIG_DIR_PLACE_HOLDER" "'$5'"
```
?

File: windows-installer/setup.nsi (hunk 5 above)
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Something's off with the indentation in these 3 lines.

File: windows-installer/setup.nsi [hunk 8]
```diff
@@ -79,8 +82,9 @@ RequestExecutionLevel admin
 ; FUNCTIONS
 ;--------------------------------
 
-; Define StrTrimNewLines function to use below.
+; Define StrTrimNewLines and StrRep functions to use below.
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Ok, now I'm a bit confused... Why do we need to do this explicitly for `StrTrimNewLines` and `StrRep`, but not for, e.g., `WordFind`? Are these lines superfluous?

Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Should you mention `StrRep` in the comment after `!include "StrFunc.nsh"`?

File: windows-installer/setup.nsi [hunk 9]
```diff
@@ -33,6 +33,9 @@
 ; Directory to install all files into.  This makes uninstall easier and safer.
 !define MAIN_INSTDIR "$INSTDIR\Main"
 
+; Directory for the user to add custom configs.
+!define CUSTOM_CONFIG_DIR "$INSTDIR\config.d"
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: What happened to factoring out the config directory? What if you decide to move `fluentd.conf` from `$INSTDIR` to somewhere else? We'd want to move the `config.d` subdirectory as well.

File: windows-installer/fluent-template.conf [hunk 10]
```diff
@@ -1,5 +1,7 @@
 # Master configuration template file for google-fluentd on windows.
 
+@include CUSTOM_CONFIG_DIR_PLACE_HOLDER
```
Comment (igorpeshansky) [inline windows-installer/fluent-template.conf]: Why not just `@include config.d/*.conf`, to make it a directory relative to the location of `fluentd.conf`? You could use a placeholder variable for `config.d`, to ensure that it doesn't fall out of sync with the code in the installer that creates it.

Comment (iantalarico) [inline windows-installer/fluent-template.conf]: Done

File: windows-installer/setup.nsi (hunk 9 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: Done

File: windows-installer/setup.nsi (hunk 8 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: I have no idea why this is what the docs says and things will fail otherwise.

Comment (iantalarico) [inline windows-installer/setup.nsi]: Done

File: windows-installer/setup.nsi (hunk 6 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: There is some weirdness with a lot of these built in functions and unicode (which is newer to NSIS and  we use) which is why I replace the whole line.

File: windows-installer/setup.nsi (hunk 5 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: Done

Comment (iantalarico) [inline windows-installer/setup.nsi]: I'd rather just use the backslash and fix fluentd.  If you feel strongly we can do that in a separate PR.

File: windows-installer/setup.nsi (hunk 7 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: See above

File: windows-installer/setup.nsi (hunk 6 above)
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Might be helpful to say exactly this in a comment...

File: windows-installer/setup.nsi (hunk 9 above)
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Not quite — see my original suggestion with `FLUENTD_CONFIG_DIRECTORY`... That way you ensure that `fluentd.conf` and `config.d` are in the same directory.

File: windows-installer/setup.nsi (hunk 6 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: It's mentioned above so I don't have to repeat it in each instance.

File: windows-installer/setup.nsi (hunk 9 above)
Comment (iantalarico) [inline windows-installer/setup.nsi]: Sorry I don't follow.  fluentd.conf and config.d are in the same directory now.

Comment (igorpeshansky) [inline windows-installer/setup.nsi]: I should have been clearer. Let's look at a concrete scenario: suppose at some future point we decide to move `fluentd.conf` under `$INSTDIR\etc`, to separate it out from other files in that directory. The way the installer code is now, whoever does it will have to remember to both change `FLUENTD_CONFIG_LOCATION` and update the `CreateDirectory` command in line 193. If they forget and only update the former, the config will break. That's why I proposed defining `FLUENTD_CONFIG_DIRECTORY` and using it in both places instead of `$INSTDIR`.

File: windows-installer/setup.nsi [hunk 11]
```diff
@@ -185,13 +189,14 @@ Section "Install"
 
   ; Create a directory to store position files.
   CreateDirectory ${MAIN_INSTDIR}\pos
+  ; Create a directory for custom configs.
+  CreateDirectory ${CUSTOM_CONFIG_DIR}
 
   ; Copy and update the fluentd config and show status, we cannot use most of
   ; the needed plugins that would do this in a better way as they do not work
   ; well with unicode on and fail (mostly silenely).
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Typo: silenely.

File: windows-installer/setup.nsi (hunk 6 above)
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: I see a reference to Unicode above, but it doesn't explain why you're replacing the whole line or what the"needed plugins" are (or what is the "better way"). Since you've mentioned Unicode already, you can just say "Replace the whole line to avoid Unicode issues" (or some such) here.

**Comment by iantalarico** [comment]
I see what you mean now, sorry.  Fixed up.  PTAL.

**Comment by iantalarico** [comment]
@igorpeshansky Friendly Ping!

File: windows-installer/setup.nsi [hunk 12]
```diff
@@ -212,15 +220,26 @@ Section "Install"
...
-    ; ever be 0 or 1.  We only have it in the template file once.
-    ${WordFind} "$3" "POS_FILE_PLACE_HOLDER" "#" $4
-
-    ; If we hit the place holder line replace it with the proper pos_file.
+	
+    ; Look for 'WIN_EVT_POS_FILE_PLACE_HOLDER', if found replace it with the
+    ; proper pos_file.
+    ${WordFind} "$3" "WIN_EVT_POS_FILE_PLACE_HOLDER" "#" $4
     ${If} $4 == "1"
+	  ; Replace the whole line instead of using "StrRep" to avoid unicode issues.
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Indentation.

File: windows-installer/setup.nsi [hunk 13]
```diff
@@ -212,15 +220,26 @@ Section "Install"
...
     ${EndIf}
 
+    ; Look for 'CUSTOM_CONFIG_DIR_PLACE_HOLDER', if found replace it with the
+    ; proper path.
+    ${WordFind} "$3" "CUSTOM_CONFIG_DIR_PLACE_HOLDER" "#" $4
+    ${If} $4 == "1"
+      ; Fluentd doesn't properly glob windows paths.  This is a temporary
+      ; fix until https://github.com/fluent/fluentd/issues/1138 is solved.
+      ${StrRep} "$5" "${CUSTOM_CONFIG_DIR}\**\*.conf" "\" "/"
+	  ; Replace the whole line instead of using "StrRep" to avoid unicode issues.
```
Comment (igorpeshansky) [inline windows-installer/setup.nsi]: Indentation.

File: windows-installer/fluent-template.conf (hunk 10 above)
Comment (igorpeshansky) [inline windows-installer/fluent-template.conf]: Nit: technically, this is not a custom config **dir**(ectory), because it ends up being a glob. How about just `CUSTOM_CONFIG_PLACE_HOLDER`?

**Comment by iantalarico** [comment]
Fixed, thanks! PTAL
